Fix sign-in for users with many groups or content or media start nodes

ondrejpialek opened the pull request:
This fixes https://github.com/umbraco/Umbraco-CMS/issues/7727

## Background

Umbraco's `BackOfficeCookieManager` inherits from a `ChunkingCookieManager` which allows authentication cookie values to grow to any size. This is useful when a user is assigned many groups, content or media starts nodes, or any other data that gets serialized into the auth cookie.

What `ChunkingCookieManager` does is quite simple - it splits values that are too long for a single cookie into multiple cookies. When you access a chunked cookie and not expect any chunking, all you will get back is what is called a header valued, e.g. "chunks:N".

## Description

There were two places that did not expect or allow chunked cookies, which is now fixed.

Firstly it was the custom `FormsAuthenticationSecureDataFormat` class, which is assigned to the `TicketDataFormat` of `UmbracoBackOfficeCookieAuthOptions`, essentially providing a string representation of the user identity. It relied on `FormsAuthentication.Encrypt` and `FormsAuthentication.Decrypt` to produce the string value.

When a user affected by this issue log in the following happened:

1. The `FormsAuthentication.Encrypt` method worked fine, it produced a very long encrypted string value
2. The long value got correctly split into chunks by the `ChunkingCookieManager`
3. Cookies were correctly sent back to Umbraco with subsequent request
4. The long cookie value was correctly reconstructed by the `ChunkingCookieManager`
5. The `FormsAuthentication.Decrypt` method failed to decrypt the value. Turns out it has an arbitrary upper limit to the input string of 4k (expecting a value from a single cookie).

To fix this problem a new class called `MachineKeyEncryptedFormsAuthSecureDataFormat` was created and is used instead of the `FormsAuthenticationSecureDataFormat`. This class replaced `FormsAuthentication.Encrypt` and `FormsAuthentication.Decrypt` with `MachineKey.Protect` and `MachineKey.Unprotect`. These methods do not have the upper limit on the encrypted string length so avoid that problem entirely. These methods work with a byte array, so a `BinaryFormatter` was used to serialize the `FormsAuthenticationTicket`. A Base64 encoding was used to turn the encrypted byte array into text.

Using the machine key for encryption has one important benefit: Umbraco Auth cookies will still work in load-balancing scenarios. In these scenarios people are used to set the same machine key on all nodes, and this encryption mechanism honours that contract, so no additional changes are needed - it will keep working.

This introduces a "non-public" breaking change in that the cookie value itself has its content formatted differently.

The second place that needed fixing was a couple of static helpers that extracted the `FormsAuthenticationTicket` from the cookie. These have been updated to properly understand chunked cookies and use the new serialization / deserialzation. Their public API did not change.

The only change that might have some impact here is that I had to add a reference to `Microsoft.Owin.Host.SystemWeb` to `Umbraco.Core`. `Umbraco.Core` already had other OWIN dll references and other project already referenced `Microsoft.Owin.Host.SystemWeb` so hopefully that is ok. The only reason this reference was added was for an extension method that allows you to obtain the OWIN context from a `HttpContentBase`. `OwinContext` is needed by the `ChunkingCookieManager` which reconstructs the chunked cookies.

## Reproducing the problem

1. Create an empty Umbraco site (or run Umbraco in the solution)
2. Create a test user
3. Create 20 user groups with a really long names and aliases (such as guids)
4. Assign these user groups to the user
5. Attempt to log in as the test user
6. Observe the login fails
7. Inspect the `PostLogin` request in browser console - observe chunked cookies were sent

When you run Umbraco with this fix is place, do the following:

1. Log in as the user - log in works
2. Inspect the `PostLogin` request in browser console - observe cookie is not chunked due to different formatting
3. Create 80 more users groups and assign them to the user
4. Log in
5. Observe log in still works
6. Inspect the `PostLogin` request in browser console - observe chunked cookies were sent

## Examples

Chunked cookie before applying the fix (login fails):
![image](https://user-images.githubusercontent.com/580850/102242815-c4d56e80-3efa-11eb-9a62-f99a868949d3.png)

Chunked cookie after applying the fix (login succeeds):
![image](https://user-images.githubusercontent.com/580850/102242616-88a20e00-3efa-11eb-9c7c-1466557b7572.png)

nul800sebastiaan commented:
Thanks @ondrejpialek! Could you please check the failing tests? https://dev.azure.com/umbraco/Umbraco%20Cms/_build/results?buildId=53593&view=ms.vss-test-web.build-test-results-tab

I had a quick look to see if it fails locally and I got it down to this error message:

```
Object reference not set to an instance of an object.
   at Umbraco.Core.Security.AuthenticationExtensions.GetUmbracoAuthTicket(HttpContextBase http) in D:\Dev\Umbraco-CMS\src\Umbraco.Core\Security\AuthenticationExtensions.cs:line 305
   at Umbraco.Web.WebApi.Filters.UmbracoUserTimeoutFilterAttribute.OnActionExecuted(HttpActionExecutedContext actionExecutedContext) in D:\Dev\Umbraco-CMS\src\Umbraco.Web\WebApi\Filters\UmbracoUserTimeoutFilterAttribute.cs:line 24
   at System.Web.Http.Filters.ActionFilterAttribute.OnActionExecutedAsync(HttpActionExecutedContext actionExecutedContext, CancellationToken cancellationToken)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at System.Web.Http.Filters.ActionFilterAttribute.<ExecuteActionFilterAsyncCore>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at System.Web.Http.Filters.ActionFilterAttribute.<ExecuteActionFilterAsyncCore>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()
```

nul800sebastiaan commented:
Just to set some expectations on this one: we won't have time this year to do a proper review of this PR, we'll try to get to it early next year. I will also aks around to see if this is actually the way we want to solve the issue, I am not super familiar with this area of the code so I'll need to rely on my colleague's advise.

ondrejpialek commented:
Thank you for the update @nul800sebastiaan and I understand reviewing this might take a while - thanks for managing my expectations!

I managed to found the issue - it revolved around the OWIN context not being properly discoverable from inside unit tests and the OWIN TestServer. Turns out the HttpContextBase is mocked out and does not have all the properties it usually has. I managed to get around that problem by locating the OWIN context directly without having to go through the HttpContextBase.

I also fixed one unit test that started failing once the OWIN context was properly discoverable, so that is a plus (details in the commit message if you are interested).

Fingers crossed for the next unit test run, if there are other issues I will try to address them as well. Sorry for not running them earlier myself - it did not work from Rider but I switched to Visual Studio and got it working.

Cheers,
Ondrej.
